### PR TITLE
Force field parameters for Poly (para-phenylene ethynylene)s

### PR DESCRIPTION
**Pull request by mbrosz**
Martini 3 force field parameter for Poly (para phenylene ethynylenes)s.

Some general remaks for the manual post-processing:

; General remarks:
; After "polyplying" PPE to the desired polymer length, some manual adjustments are recommended:
; 1) PPE ends with an aromatic ring and not with a triple bond: Delete last bead of topology and coordinate file and all respective potentials using that bead. 
; 2) To prevent bending over of last and first aromatic ring: Add an additional bond angle potential at both polymer ends. (k=50,phi0=165)
;    (e.g. bond angle potential between 1-4-8 and between End-(End-4)-(End-8) ).
; 3) For furher explanation, please look at https://doi.org/10.1039/D1CP04237H.

I added these also in the PPE.martini3.ff file

### CODE REVIEW

**Comment by fgrunewald** [comment]
@mbrosz thanks for preparing this PR. I've some suggestions on how to improve the parameter generation based on your comments in the .ff file. 

- you add your paper into the citations.bib file using the BIBTEX syntax with the '{'. Essentially copy the citation from google scholar and add it there. Then you can add the keyword to the [citations] directive and your citation will be automatically added
- if I understand point 1 correctly your last residue does not have the same number of beads as the other residues, but rather it ends in a BENZ residue, right? I would suggest you define a block called PPEter that is only for the terminal residue. Then people have to specify `-seq PPE:n_mon PPEter:1` to get the correct itp file for PPE. It is less intuitive but least they don't have to manually go into the file. For the future we are working on a syntax/implementation that will take care of this issue automatically. 
- when you have defined the PPEter block you can also define an additional link that will automatically add the angle to the itp file in the same way that you did now. Be aware that link resname should be defined as `resname "PPE|PPEter" `
- you can also add a test for the itp file generation that will run automatically and compare to a reference itp. To do that have a look in `polyply/tests/test_data/library_tests/martini3/`. Add a folder called PPE and the add a folder called polyply. Within that folder you place the reference itp and a command file. The command file contains the polyply command. See the other directories as an example. Once you have done that edit the file `polyply/tests/test_lib_files.py` and add PEE in line 176ff by adding `['martini3', 'PPE']`. Commit and done. 

Let me know if you need any help with the above suggestions.

**Comment by mbrosz** [comment]
I double checked the results from the polyply output, with mine. 
Now I hope the implementation of the PPEs within the polylib should be fine. 
The final files I have just uploaded. So maye be just to triple check, another short test run. 
Cheers,
Matthias

File: polyply/data/martini3/PPE.martini3.ff [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    
+[ angles ]
+   1    3    4     1    143    550  	{"comment": "angle aromatic ring-triple bond"}
+   2    3    4     1     83    650   	{"comment": "angle aromatic ring-triple bond"}
+   
+[ dihedrals ]
+    4    2    1   3    2  0    50   	{"comment": "improper dihedral left turn"}
+    
+[ link ]
+resname "PPE|PPEter"
```
Comment (fgrunewald) [inline polyply/data/martini3/PPE.martini3.ff:41]: This link only concerns PPE and not PPEter or am I mistaken? If so then you can remove the second residue name

File: polyply/tests/test_data/library_tests/martini3/PPE/polyply/PPE.itp [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+77    78    1 0.32500 
+77    79    1 0.32500 
+78    79    1 0.32500 
+81    82    1 0.32500 
+81    83    1 0.32500 
+82    83    1 0.32500 
+85    86    1 0.32500 
+85    87    1 0.32500 
+86    87    1 0.32500 
+#ifdef POSRES
```
Comment (fgrunewald) [inline polyply/tests/test_data/library_tests/martini3/PPE/polyply/PPE.itp]: For the automatic check, I think we cannot deal with position_restraints that have this force constant syntax. Hence the fail of the test. Can you remove this section that deals with the positions_restraints. 

Also are they needed for equilibriation etc.? If so I think we could also implement them in polyply by adding position restraints to the ff-file blocks. However, then you cannot add the #define directive automatically. Not sure that would help then.

**Review by fgrunewald — CHANGES_REQUESTED** [review]
Hi Matthias, 

Thanks for adding all the files. Overall you did a great job and don't worry about the GitHub stuff. I made some specific comments that need addressing. I'm not sure why you included two itp files with the tests. In principle one itp file should be enough. I assume you commited the other file by accident. In that case just remove it and commit the deleted file. 

In order to test locally if all the tests work you can do the following: Go to the polyply directory on your machine and run `pip install -r requirements-tests.txt`. This will install all packages needed for testing. Then you can go to the test folder and for example run `pytest test_lib_files.py`. This should run all integration tests. So you can play around a little with making it work. Sometimes it takes two three iterations to get rid of some problems. Don't be discouraged by that. 

Finally, I would ask you to add PPE, PEEinit, PPEter all into one file called PPE.martini3.ff. This is just for bookkeeping. They don't have to be in separate files. 

Let me know if you encounter any problems with the testing. I also enable the automated CI, which means every-time you push to this branch the full testing suite will be run. If all is good at the bottom everything should be green. 

Cheers,
Fabian

File: polyply/tests/test_data/library_tests/martini3/PPE/polyply/PPE.itp [hunk 3]
```diff
@@ -0,0 +1,349 @@
+[   moleculetype   ]
+PPE   4
```
[the remark this answers is not in the document]
Comment (mbrosz) [reply polyply/tests/test_data/library_tests/martini3/PPE/polyply/PPE.itp]: correct, I adjusted the value to 3

File: polyply/tests/test_data/library_tests/martini3/PPE/polyply/PPE_20mers.itp [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+77    78    1 0.32500 
+77    79    1 0.32500 
+78    79    1 0.32500 
+81    82    1 0.32500 
+81    83    1 0.32500 
+82    83    1 0.32500 
+85    86    1 0.32500 
+85    87    1 0.32500 
+86    87    1 0.32500 
+#ifdef POSRES
```
[the remark this answers is not in the document]
Comment (mbrosz) [reply polyply/tests/test_data/library_tests/martini3/PPE/polyply/PPE_20mers.itp]: check

File: polyply/tests/test_data/library_tests/martini3/PPE/polyply/PPE.itp (hunk 2 above)
Comment (mbrosz) [reply polyply/tests/test_data/library_tests/martini3/PPE/polyply/PPE.itp]: check

File: polyply/data/martini3/citations.bib [hunk 5]
```diff
@@ -19,3 +19,16 @@ @article{polyply
...
...stiffness. Due to this, PPEs have a wide range of technological applications. Although the material properties of single-chains or mixtures of few PPE chains have been studied in detail, the properties of large assemblies remain to be fully explored. Here, we developed a coarse-grained model for PPEs with the Martini 3 force field to enable computational studies of PPEs in large-scale assembly. We used an optimization geometrical approach to take the shape of the π conjugated backbone into account and also applied an additional angular potential to tune the mechanical bending stiffness of the polymer. Our Martini 3 model reproduces key structural and thermodynamic observables of single PPE chains and mixtures, such as persistence length, density, packing and stacking. We show that chain entanglement increases with the expense of nematic ordering with growing PPE chain length. With the Martini 3 PPE model at hand, we are now able to cover large spatio-temporal scales and thereby to uncover key aspects for the structural organization of PPE bulk systems. The model is also predicted to be of high applicability to investigate out-of-equilibrium behavior of PPEs under mechanical force.}
```
[the remark this answers is not in the document]
Comment (mbrosz) [reply polyply/data/martini3/citations.bib]: abstract removed

**Review by fgrunewald — COMMENTED** [review]
Seems the itps produced by polyply and the reference are not the same

**Review by fgrunewald — COMMENTED** [review]
@mbrosz looks good to me. It also seems the integration test with your reference file are now passing. Is there anything else you would like to check before merging? Perhaps it would be good to see if the structure generation works right away. Did you test that?

**Comment by mbrosz** [comment]
Hi Fabian, 
kind of took me a while to find out all the problems with the script, however, the test_lib.py worked pretty well and I could identify my issues very fast. Thanks for the constant help during the pull request. 
I also tried the polyply gen_coords command to generate the out.gro file, which also worked for me. In addition, I performed some small test run and this also worked, so I assume  everything should be fine now.

**Review by fgrunewald — APPROVED** [review]
@mbrosz Good to hear that the library tool was useful. Considering the tests you did I agree that everything is in place and we can merge it. Perhaps we should add some usage notes to the wiki and of course tweet it out, so people are aware. Do you have a Twitter handle?

**Comment by mbrosz** [comment]
Wonderful, I think a short comment about in the wiki might be useful. I can think about some basic instructions to build the model with polyply and write you something, if you prefer it this way. However, might be good to get some example article to see how the overall layout about the wiki reports look like?
The tweeting out would be a very cool idea I think. Since I am actually not on twitter, I would suggest to use the handle from out group: @graeter_mbm 
I hope this is also fine with you.

**Comment by fgrunewald** [comment]
@mbrosz I've no strong idea about how this should look like. As far as I am concerned it can be a bunch of bullet points saying what to watch out for. For example, in order to generate an itp you need to also specify the termini etc. I will now merge this PR and create a new issue for this, then we see. Thanks for the twitter handle. I'll tweet something next week when I also released a new version to pypi. Thanks for your contribution and I hope many people will use this polymer in their studies.

**Pull request merged**